### PR TITLE
feat: Implement SQL Identity Hub Store

### LINKED ISSUE #38: Persistent IdentityHubStore - PostgreSQL extension
# Feature Request
Child issue of #29 

## Solution Proposal
Many examples available in the code base, e.g. https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/tree/main/extensions/control-plane/store/sql/asset-index-sql

## Type of Issue
Improvement.

## Checklist

- [ ] assigned appropriate label?
- [x] **Do NOT select a milestone or an assignee!**

### PR DESCRIPTION
**Pull request by bscholtes1A**
## What this PR changes/adds

Provides SQL implementation of `IdentityHubStore`.

## Why it does that

Ensuring persistence of items put in the store.

## Linked Issue(s)

Closes #38.

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/store/sql/identity-hub-store-sql/docs/schema.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE IF NOT EXISTS edc_identityhub
+(
+    item  VARCHAR NOT NULL UNIQUE
```
Comment (paullatzelsperger) [inline extensions/store/sql/identity-hub-store-sql/docs/schema.sql:3]: I'm not quite sure about the schema to be honest. As I've alluded to in [this comment](https://github.com/eclipse-dataspaceconnector/IdentityHub/pull/43#discussion_r1000817845), even though IH objects are "schema-less", I would prefer to create some sort of `IdentityHubRecord` class that wraps the hub object, and provides things like an ID, date of entry, date of last update, etc.
For reference, EDC already provides the `MutableEntity` class, which could be extended here.
<br/>
Also, even though the `item` col is constrained to `NOT NULL UNIQUE`, it is not a primary key, which makes it difficult to search for and maintain the entries, for example Issue #45 will add some methods to the IH interface.
A PK is also `NOT NULL UNIQUE`, but in addition, it is automatically indexed.
<br/>
Lastly, once the `IdentityHubRecord` object is in place, I'd rename the data column to something like `payload` or `content`.

File: extensions/store/sql/identity-hub-store-sql/src/main/java/org/eclipse/dataspaceconnector/identityhub/store/sql/SqlIdentityHubStore.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    @Override
+    public Collection<byte[]> getAll() {
+        return transactionContext.execute(() -> {
+            try (var connection = getConnection()) {
+                try (var stream = executeQuery(connection, true, this::parse, identityHubStatements.getSelectAllItemsTemplate())) {
+                    return stream.collect(Collectors.toList());
+                }
+            } catch (EdcPersistenceException e) {
```
Comment (paullatzelsperger) [inline extensions/store/sql/identity-hub-store-sql/src/main/java/org/eclipse/dataspaceconnector/identityhub/store/sql/SqlIdentityHubStore.java:61]: necessary?

File: extensions/store/sql/identity-hub-store-sql/src/main/java/org/eclipse/dataspaceconnector/identityhub/store/sql/SqlIdentityHubStore.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            }
+        });
+    }
+
+    @Override
+    public void add(byte[] hubObject) {
+        transactionContext.execute(() -> {
+            try (var connection = getConnection()) {
+                executeQuery(connection, identityHubStatements.getInsertItemTemplate(), new String(hubObject));
+            } catch (EdcPersistenceException e) {
```
Comment (paullatzelsperger) [inline extensions/store/sql/identity-hub-store-sql/src/main/java/org/eclipse/dataspaceconnector/identityhub/store/sql/SqlIdentityHubStore.java:74]: necessary?

File: spi/identity-hub-store-spi/src/testFixtures/java/org/eclipse/dataspaceconnector/identityhub/store/spi/IdentityHubStoreTestBase.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public abstract class IdentityHubStoreTestBase {
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+    private static final Random RAND = new Random();
+
+    @Test
+    void saveAndListVerifiableCredentials() {
```
Comment (paullatzelsperger) [inline spi/identity-hub-store-spi/src/testFixtures/java/org/eclipse/dataspaceconnector/identityhub/store/spi/IdentityHubStoreTestBase.java:36]: The test coverage could be extended, eg.: 
- getAll_whenNothingInDb
- add_whenExists

**Review by wolf4ood — APPROVED** [review]

File: extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/api/IdentityHubControllerTest.java [hunk 5]
```diff
@@ -151,6 +152,8 @@ private RequestObject createRequestObject(String method, byte[] data) {
                                 .descriptor(Descriptor.Builder.newInstance()
                                         .method(method)
                                         .nonce(NONCE)
+                                        .recordId(UUID.randomUUID().toString())
+                                        .dateCreated(Instant.now().getEpochSecond())
```
Comment (jimmarino) [inline extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/api/IdentityHubControllerTest.java:156]: Should this be millis or is there a specific reason why this is the epoch date?

**Review by jimmarino — APPROVED** [review]

**Comment by jimmarino** [comment]
@bscholtes1A Paul is out, so I took a look. I had one small comment - do you want to review that and also fix the conflicts so we can merge?

File: extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/api/IdentityHubControllerTest.java (hunk 5 above)
Comment (bscholtes1A) [reply extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/api/IdentityHubControllerTest.java:156]: From what I understand from the Decentralized Identity Foundation it should be epoch seconds: https://identity.foundation/decentralized-web-node/spec/0.0.1-predraft/#write

> The object MUST contain a dateCreated property, and its value MUST be an [Unix epoch timestamp](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_16) that MUST be set and interpreted as the time the logical entry was created by the DID owner or another permitted party.

**Comment by bscholtes1A** [comment]
> @bscholtes1A Paul is out, so I took a look. I had one small comment - do you want to review that and also fix the conflicts so we can merge?

Answered and conflict fixed.

File: extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/api/IdentityHubControllerTest.java (hunk 5 above)
Comment (jimmarino) [reply extensions/identity-hub/src/test/java/org/eclipse/dataspaceconnector/identityhub/api/IdentityHubControllerTest.java:156]: OK thx

**Comment by jimmarino** [comment]
@bscholtes1A do you want to merge?

**Pull request merged**